cosmo/types.py

```
@@ -0,0 +1,433 @@
+
+    def getIPInterfaceObject(self) -> IPv4Interface|IPv6Interface:
+        return ipaddress.ip_interface(self.getIPAddress())
+
+    def isGlobal(self) -> bool:
+        return self.getIPInterfaceObject().is_global
+
+    def getRole(self) -> str:
+        if "role" in self:
+            return self["role"]
+
+
+class TagType(AbstractNetboxType):
+    _delimiter = ':'
+    def getTagComponents(self):
+        return self['name'].split(self._delimiter)
+    def getTagName(self):
+        return self.getTagComponents()[0]
+    def getTagValue(self):
+        return self.getTagComponents()[1]
+
+
+class RouteTargetType(AbstractNetboxType):
+    # RT is present in VRFs
+    def getName(self):
+        return self["name"]
+
+
+class VRFType(AbstractNetboxType):
+    def getName(self) -> str:
+        return self["name"]
+
+    def getDescription(self) -> str:
+        return self["description"]
+
+    def getExportTargets(self) -> list[RouteTargetType]:
+        return self["export_targets"]
+
+    def getImportTargets(self) -> list[RouteTargetType]:
+        return self["import_targets"]
+
+    def getRouteDistinguisher(self) -> str:
+        return self["rd"]
+
+
+class InterfaceType(AbstractNetboxType):
+    def __repr__(self):
+        return super().__repr__() + f"({self.getName()})"
+
+    def getName(self) -> str:
+        return self['name']
+
+    def getMACAddress(self) -> str|None:
+        if "mac_address" in self:
+            return self['mac_address']
+
+    def getUntaggedVLAN(self):
+        cf = self.getCustomFields()
+        if "untagged_vlan" in self.keys() and self["untagged_vlan"]:
+            return self["untagged_vlan"]
+        elif "outer_tag" in cf.keys() and cf["outer_tag"]:
+            # outer_tag should not be used with untagged vlan
+            # untagged_vlan has priority over outer_tag
+            # we have to build the VLANType object in the case of
+            # outer_tag usage, since there's no native type in netbox
```

Let's throw an error here, if `outer_tag` and `untagged_vlan` is used together.

---

done!